segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/S3ClientWrapper.java

```diff
@@ -0,0 +1,262 @@
+import java.io.ByteArrayInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import lombok.SneakyThrows;
+
+/**
+ * Client wrapper for S3JerseyClient.
+ */
+class S3ClientWrapper extends S3JerseyClient {
+    private final ConcurrentMap<String, AclSize> aclMap;
+    private final String baseDir;
+    private final ConcurrentMap<String, ConcurrentMap<Integer, CopyPartRequest>> multipartUploads = new ConcurrentHashMap<>();
+
+    public S3ClientWrapper(S3Config exts3Config, ConcurrentMap<String, AclSize> aclMap, String baseDir) {
+        super(exts3Config);
+        this.aclMap = aclMap;
+        this.baseDir = baseDir;
+    }
+
+    @Override
+    public DeleteObjectsResult deleteObjects(DeleteObjectsRequest request) {
+        return new DeleteObjectsResult();
+    }
+
+    @Override
+    public PutObjectResult putObject(PutObjectRequest request) {
+
+        synchronized(this.aclMap) {
+            if (request.getObjectMetadata() != null) {
+                request.setObjectMetadata(null);
+            }
+            try {
+                Path path = Paths.get(this.baseDir, request.getBucketName(), request.getKey());
+                Files.createDirectories(path.getParent());
+                Files.createFile(path);
+            } catch (IOException e) {
+                throw new S3Exception(e.getMessage(), 0);
+            }
+            PutObjectResult retVal = new PutObjectResult();
+            if (request.getAcl() != null) {
+                long size = 0;
+                if (request.getRange() != null) {
+                    size = request.getRange().getLast() + 1;
+                }
+                aclMap.putIfAbsent(request.getKey(), new AclSize(request.getAcl(),
+                        size));
+            }
+            return retVal;
+        }
+    }
+
+    @Override
+    @SneakyThrows(BadOffsetException.class)
+    public void putObject(String bucketName, String key, Range range, Object content) {
+
+        Path path = Paths.get(this.baseDir, bucketName, key);
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE)) {
+
+            long bytesTransferred = channel.transferFrom(Channels.newChannel((InputStream) content),
+                    range.getFirst(), range.getLast() + 1 - range.getFirst());
+
+            AclSize aclKey = aclMap.get(key);
+            aclMap.put(key, aclKey.withSize(range.getLast() + 1));
+        } catch (IOException | IllegalArgumentException e) {
+            throw new BadOffsetException(key, 0, 0);
```

Comment:
It is not obvious that all IOExceptions and IllegalArguments are the result of a badOffset: "Exception that is thrown whenever a Write failed due to a bad offset". Additionally 0 and 0 are clearly incorrect values.